File: node-api/src/specs/matrixSpec.js
/**
 * @file matrixSpec.js
 * @description SDD Spec — defines the data contracts for the matrix statistics workflow.
 *
 * In Spec-Driven Development, specs act as the source of truth for what
 * data enters and exits each agent. Agents are built to satisfy specs,
 * and tests validate against specs.
 */

/**
 * @typedef {Object} QRResult
 * @property {number[][]} Q  - Orthogonal factor matrix
 * @property {number[][]} R  - Upper-triangular factor matrix
 * @property {number} originalRows
 * @property {number} originalCols
 */

/**
 * @typedef {Object} StatsRequest
 * @property {QRResult} qrResult - Matrices received from Go API
 */

/**
 * @typedef {Object} DiagonalCheck
 * @property {boolean} Q - Whether Q is diagonal
 * @property {boolean} R - Whether R is diagonal
 */

/**
 * @typedef {Object} StatsResponse
 * @property {number} max          - Maximum value across both matrices
 * @property {number} min          - Minimum value across both matrices
 * @property {number} average      - Average of all values
 * @property {number} sum          - Sum of all values
 * @property {DiagonalCheck} isDiagonal - Diagonal check per matrix
 */

/**
 * Validates that a StatsRequest has the required shape.
 * @param {unknown} body
 * @returns {{ valid: boolean, error?: string }}
 */
function validateStatsRequest(body) {
  if (!body || typeof body !== "object") {
    return { valid: false, error: "Request body must be an object" };
  }
  const { qrResult } = body;
  if (!qrResult) {
    return { valid: false, error: "Missing field: qrResult" };
  }
  if (!Array.isArray(qrResult.Q) || !Array.isArray(qrResult.R)) {
    return { valid: false, error: "qrResult.Q and qrResult.R must be arrays" };
  }
  if (qrResult.Q.length === 0 || qrResult.R.length === 0) {
    return { valid: false, error: "Q and R matrices must not be empty" };
  }
  return { valid: true };
}

module.exports = { validateStatsRequest };

File: go-api/internal/specs/matrix_spec.go
// Package specs defines the data contracts (specs) used by the SDD agents.
// Each spec describes the input/output shape of an operation so that
// agents can be composed and validated independently.
package specs

// MatrixInput is the expected request body for the QR endpoint.
type MatrixInput struct {
	// Matrix is a rectangular array of float64 values.
	// Rows must be >= Cols for QR to be well-defined.
	Matrix [][]float64 `json:"matrix" validate:"required,min=1"`
}

// QRResult holds the two factor matrices returned by QR decomposition.
type QRResult struct {
	// Q is an orthogonal matrix (rows x rows)
	Q [][]float64 `json:"Q"`
	// R is an upper-triangular matrix (rows x cols)
	R [][]float64 `json:"R"`
	// OriginalRows is the number of rows in the input matrix
	OriginalRows int `json:"originalRows"`
	// OriginalCols is the number of columns in the input matrix
	OriginalCols int `json:"originalCols"`
}

// NodeAPIRequest is the payload sent from Go API → Node.js API.
type NodeAPIRequest struct {
	QRResult QRResult `json:"qrResult"`
}

// TokenRequest is used for JWT generation.
type TokenRequest struct {
	Username string `json:"username"`
	Password string `json:"password"`
}

// TokenResponse returns a signed JWT.
type TokenResponse struct {
	Token string `json:"token"`
}
